lib/Pool.js

```diff
@@ -247,6 +247,29 @@ Pool.prototype._removeConnection = function(connection) {
   this.releaseConnection(connection);
 };
 
+Pool.prototype.restoreUser = function restoreUser(connection, callback) {
+  var poolConfig = this.config.connectionConfig;
+  var pool = this;
+
+  var config = connection.config;
+  var user = poolConfig.user;
+  var database = poolConfig.database;
+  var password = poolConfig.password;
+
+  // only when credentials have been changed from the pool defaults
+  if (config.user !== user || config.database !== database || config.password !== password) {
+    connection.changeUser({ user: user, database: database, password: password }, function (err) {
+      if (err) return callback(err);
+      // emit as if it were a new connection so consumers can setup session state again
+      pool.emit('connection', connection);
+      callback(null);
+    });
+    connection._purge = false;
+  } else {
+    callback();
```

Comment:
This callback needs to be in a `process.nextTick` so that `restoreUser` is always async instead of sometimes async.